### PR TITLE
Add detector type to config

### PR DESCRIPTION
**Pull request by mdevino**
This addresses #165.

I'm opening this as I'd like to get some feedback on what I've done so far.

Missing tasks:
* [x] Add ADR
* [x] Either use generics or create a helper function for `text_contents(), `text_generation()` and `text_context_doc()` - as [faux has a limitation when dealing with generics](https://github.com/nrxus/faux/issues/18), I've rolled back this change and will work on it in a different PR.

### CODE REVIEW

File: src/clients.rs [hunk 1]
```diff
@@ -129,16 +129,31 @@ impl From<tonic::Status> for Error {
 pub struct HttpClient {
     base_url: Url,
     client: reqwest::Client,
+    detector_type: DetectorType,
```
Comment (mdevino) [inline src/clients.rs:132]: Is this okay, or should we have a "subclass" for detector clients?

File: src/clients.rs [hunk 2]
```diff
@@ -129,16 +129,31 @@ impl From<tonic::Status> for Error {
...
+            client,
+            detector_type,
+        }
     }
 
     pub fn base_url(&self) -> &Url {
         &self.base_url
     }
+
+    pub fn endpoint(&self) -> String {
```
Comment (mdevino) [inline src/clients.rs:148]: Would it be better to return `String`, `&'static str` or `&Url` here?

File: src/clients.rs [hunk 3]
```diff
@@ -129,16 +129,31 @@ impl From<tonic::Status> for Error {
...
+        let path = match self.detector_type {
+            DetectorType::Content => "/api/v1/text/contents",
+            DetectorType::Generated => "/api/v1/text/generation",
+            DetectorType::Chat => "/api/v1/text/context/chat",
+            DetectorType::Context => "/api/v1/text/context/doc",
+        };
+        self.base_url.join(path).unwrap().to_string()
+    }
```
Comment (mdevino) [inline src/clients.rs:156]: Would it be better to return `String`, `&'static str` or `&Url` here?

File: src/config.rs [hunk 4]
```diff
@@ -119,6 +119,17 @@ pub struct ChunkerConfig {
     pub service: ServiceConfig,
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, Default)]
+#[cfg_attr(test, derive(PartialEq))]
+#[serde(rename_all = "lowercase")]
+pub enum DetectorType {
+    #[default]
```
Comment (mdevino) [inline src/config.rs:126]: Note: even though this is set as default, "type" is mandatory on the YAML.

**Review by declark1 — COMMENTED** [review]
We should not add detector concepts, e.g. `detector_type` and `endpoint` with detector-specific endpoints to `HttpClient`. This is meant to be a reusable abstraction for any HTTP API, e.g. we will be using it for the upcoming OpenAI chat & chat completions implementation.

**Comment by mdevino** [comment]
> We should not add detector concepts, e.g. `detector_type` and `endpoint` with detector-specific endpoints to `HttpClient`. This is meant to be a reusable abstraction for any HTTP API, e.g. we will be using it for the upcoming OpenAI chat & chat completions implementation.

@declark1 Agreed, but not quite sure on how to go about this in a _Rusty_ way. Should I move these out to a trait, or create a whole new struct?

**Comment by declark1** [comment]
Closing this as we are consolidating client-related changes into #220

**Pull request closed without merging**